File: ant-sdk/antd-mcp/src/antd_mcp/discover.py
"""Port discovery for the antd daemon.

The antd daemon writes a ``daemon.port`` file on startup containing up to three
lines:
  - Line 1: REST port
  - Line 2: gRPC port
  - Line 3: PID of the daemon process (optional)

When line 3 is present, this module validates that the process is still alive.
If the PID refers to a dead process the port file is considered stale and
discovery returns empty results.

This module reads that file using platform-specific data directory paths to
auto-discover the daemon without requiring manual configuration.
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

_PORT_FILE_NAME = "daemon.port"
_DATA_DIR_NAME = "ant"
_SDK_SUBDIR_NAME = "sdk"


def _data_dir() -> Path | None:
    """Return the platform-specific data directory for the antd SDK daemon, or None.

    The ``sdk`` subdirectory keeps antd's port file separate from the
    ant-node daemon, which writes to the same ``ant`` umbrella dir.
    """
    if sys.platform == "win32":
        appdata = os.environ.get("APPDATA")
        if not appdata:
            return None
        return Path(appdata) / _DATA_DIR_NAME / _SDK_SUBDIR_NAME

    if sys.platform == "darwin":
        home = os.environ.get("HOME")
        if not home:
            return None
        return Path(home) / "Library" / "Application Support" / _DATA_DIR_NAME / _SDK_SUBDIR_NAME

    # Linux and other Unix-likes
    xdg = os.environ.get("XDG_DATA_HOME")
    if xdg:
        return Path(xdg) / _DATA_DIR_NAME / _SDK_SUBDIR_NAME
    home = os.environ.get("HOME")
    if not home:
        return None
    return Path(home) / ".local" / "share" / _DATA_DIR_NAME / _SDK_SUBDIR_NAME


def _is_pid_alive(pid: int) -> bool:
    """Check whether a process with the given PID is still running.

    Uses ``os.kill(pid, 0)`` which works cross-platform on Python 3.
    """
    try:
        os.kill(pid, 0)
    except ProcessLookupError:
        return False
    except PermissionError:
        # Process exists but we lack permission to signal it.
        return True
    except OSError:
        # Unable to determine — assume alive to avoid false negatives.
        return True
    return True


def _read_port_file() -> tuple[int, int]:
    """Read the daemon.port file and return (rest_port, grpc_port).

    Returns (0, 0) if the file is missing, unreadable, or stale (the
    recorded PID no longer refers to a running process).
    A single-line file is valid; grpc_port will be 0 in that case.
    """
    data_dir = _data_dir()
    if data_dir is None:
        return 0, 0

    port_file = data_dir / _PORT_FILE_NAME
    try:
        text = port_file.read_text().strip()
    except (OSError, ValueError):
        return 0, 0

    lines = text.splitlines()
    if not lines:
        return 0, 0

    # Line 3 (optional): PID of the daemon process.
    if len(lines) >= 3:
        try:
            pid = int(lines[2].strip())
        except ValueError:
            pid = None
        if pid is not None and not _is_pid_alive(pid):
            return 0, 0

    rest_port = _parse_port(lines[0])
    grpc_port = _parse_port(lines[1]) if len(lines) >= 2 else 0
    return rest_port, grpc_port


def _parse_port(s: str) -> int:
    """Parse a port string, returning 0 on failure."""
    try:
        n = int(s.strip())
        if 1 <= n <= 65535:
            return n
    except ValueError:
        pass
    return 0


def discover_daemon_url() -> str:
    """Read the daemon.port file and return the REST base URL.

    Returns ``"http://127.0.0.1:{port}"`` on success, or ``""`` if the port
    file is not found or unreadable.
    """
    rest, _ = _read_port_file()
    if rest == 0:
        return ""
    return f"http://127.0.0.1:{rest}"


def discover_grpc_target() -> str:
    """Read the daemon.port file and return the gRPC target.

    Returns ``"127.0.0.1:{port}"`` on success, or ``""`` if the port file
    is not found or has no gRPC line.
    """
    _, grpc = _read_port_file()
    if grpc == 0:
        return ""
    return f"127.0.0.1:{grpc}"

File: ant-sdk/antd-go/discover.go
package antd

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
)

const portFileName = "daemon.port"
const dataDirName = "ant"
const sdkSubDirName = "sdk"

// DiscoverDaemonURL reads the daemon.port file written by antd on startup
// and returns the REST base URL (e.g. "http://127.0.0.1:8082").
// Returns empty string if the port file is not found or unreadable.
func DiscoverDaemonURL() string {
	rest, _ := readPortFile()
	if rest == 0 {
		return ""
	}
	return fmt.Sprintf("http://127.0.0.1:%d", rest)
}

// DiscoverGrpcTarget reads the daemon.port file written by antd on startup
// and returns the gRPC target (e.g. "127.0.0.1:50051").
// Returns empty string if the port file is not found or has no gRPC line.
func DiscoverGrpcTarget() string {
	_, grpc := readPortFile()
	if grpc == 0 {
		return ""
	}
	return fmt.Sprintf("127.0.0.1:%d", grpc)
}

// readPortFile reads the daemon.port file and returns the REST and gRPC ports.
// The file format is: REST port (line 1), gRPC port (line 2), PID (line 3).
// A single-line file is valid (gRPC port will be 0).
// If a PID is present and the process is not running, the file is considered
// stale and both ports are returned as 0.
func readPortFile() (restPort, grpcPort uint16) {
	dir := dataDir()
	if dir == "" {
		return 0, 0
	}

	data, err := os.ReadFile(filepath.Join(dir, portFileName))
	if err != nil {
		return 0, 0
	}

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	if len(lines) < 1 {
		return 0, 0
	}

	// Check PID on line 3 — if present and process is dead, file is stale
	if len(lines) >= 3 {
		if pid, err := strconv.Atoi(strings.TrimSpace(lines[2])); err == nil && pid > 0 {
			if !processAlive(pid) {
				return 0, 0
			}
		}
	}

	restPort = parsePort(lines[0])
	if len(lines) >= 2 {
		grpcPort = parsePort(lines[1])
	}
	return restPort, grpcPort
}

// processAlive is implemented per-platform in discover_unix.go and discover_windows.go.

func parsePort(s string) uint16 {
	n, err := strconv.ParseUint(strings.TrimSpace(s), 10, 16)
	if err != nil {
		return 0
	}
	return uint16(n)
}

// dataDir returns the platform-specific data directory for the antd SDK daemon.
//   - Windows: %APPDATA%\ant\sdk
//   - macOS:   ~/Library/Application Support/ant/sdk
//   - Linux:   $XDG_DATA_HOME/ant/sdk or ~/.local/share/ant/sdk
//
// The sdk subdirectory keeps antd's port file separate from the ant-node
// daemon, which writes to the same ant umbrella dir.
func dataDir() string {
	switch runtime.GOOS {
	case "windows":
		appdata := os.Getenv("APPDATA")
		if appdata == "" {
			return ""
		}
		return filepath.Join(appdata, dataDirName, sdkSubDirName)

	case "darwin":
		home := os.Getenv("HOME")
		if home == "" {
			return ""
		}
		return filepath.Join(home, "Library", "Application Support", dataDirName, sdkSubDirName)

	default: // linux and others
		if xdg := os.Getenv("XDG_DATA_HOME"); xdg != "" {
			return filepath.Join(xdg, dataDirName, sdkSubDirName)
		}
		home := os.Getenv("HOME")
		if home == "" {
			return ""
		}
		return filepath.Join(home, ".local", "share", dataDirName, sdkSubDirName)
	}
}
